### PR TITLE
command foundation (for real this time)

### PR DESCRIPTION
**Pull request by djdv**
This is a (late) follow-up to https://github.com/djdv/go-filesystem-utils/pull/6#issuecomment-1177034987

**Recap:**
Rather than trying to accommodate external libraries from which we originate, it's been decided to just drop it outright and start from scratch on these components.
(Honestly this should have been done in the first place and probably would have saved a lot of time and effort. But it's easier to see that now in hindsight. The opposite was assumed prior, but ended up being a massive timesink with little gain.)

**Concept:**
We really do not need any of the complex features offered by the old cmds-lib, so the API here should be enough and has been seeing real use with our prototypes, without any real issues at all.
If we do need something, adding it shouldn't be too difficult.
The use of generics and the way things are automated implicitly has made it relatively simple to make very particular commands, without too much duplicated code for the caller.
(This is less a matter of convenience, and more a matter of consistency/correctness. But the former is apparent too.)

**Implementation:**
Some of the generic declaration syntax is horrid, but caller-side I think it's good enough.
As a general trend, I'm okay with keeping this pattern of prioritizing convenience for the caller over all else.
Even if it means spending some time occasionally in type-parameter/template hell deeper within APIs.

If something seems obviously bad, now is the time to pick at it.
The interface should be enough for what we generally need.
```go
Command interface {
	Name() string
	Synopsis() string
	Usage() string
	Subcommands() []Command
	Execute(ctx context.Context, args ...string) error
}
```
And the generic constructor/wrapper should make authoring commands simple.
```go
command.MakeCommand(
	cmdName, synopsis, usage,
	execute,
	command.WithSubcommands(subcommands...),
)
func execute(context.Context, *settings, ...string) error {
```
Where `*settings` is a loosely constrained, but otherwise arbitrary struct type.
If you can define this function and fill in the metadata, you can have and chain commands in a hierarchy.
`fs`, `fs daemon`, `fs daemon shutdown`, `fs daemon -maddr=...`, `fs something with args` `fs something with args -and="flags"`, etc.

**TODO:**
The current code seems to work, but needs more testing, refinement, review, documenting, and all the other usual things.
@byfw will be taking responsibility for advancing this, while I'm here for support and the final passes.
I'll mainly be focusing on the future components that are based on this branch, but will try not to delay feedback here for too long.

When everything seems okay, we'll squash and merge.

### CODE REVIEW

File: internal/command/command_test.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+func hlprHelpRequested(t *testing.T, hf *command.HelpFlag, want bool) {
+	t.Helper()
+	if got := hf.HelpRequested(); got != want {
+		hlprGotWant(t, "HelpFlag returned unexpected value:", got, want)
+	}
+}
+
+func hlprGotWant(t *testing.T, explain string, got, want any) {
```
Comment (byfw) [inline internal/command/command_test.go:112]: I think the tests could be simplified by making this function return an error and checking the comparison here instead.

**Comment by djdv** [comment]
Looks like I forgot to add the GitHub actions files. Sometime later I'll double check the yaml we had, and patch it into this branch to at least run the Go tests.
We don't have something for releases yet, but we also don't need it right this second either so that can probably come in with the daemon code.

***

Unimportant aside, we're floored at Go 1.18 but the daemon code was already taking advantage of these https://go.dev/doc/go1.19#atomic_types prior to release, so the version can remain 1.18 here, but will bump almost immediately after.

**Comment by djdv** [comment]
Recap from yesterday's offline discussion.

The option `WithoutArguments` is weird.
The compiler should allow us to pass a typeset of 2 different function signatures, and just know at compile time if it does or does not accept arguments.
This would make the `WithoutArguments` option implicit, internally.

This can work, but it seems to break automatic type inference.
Which causes calls like this: 
`command.MakeCommand(name, synopis, usage, execFn, options...)`
to become
`command.MakeCommand[*settingsType](name, synopis, usage, execFn, options...)`
(Which is redundant since the type parameter for `*settingsType` could be inferred from `execFn`'s parameters)

I'll see if there's an easy way to avoid this / alternative method of handling this.
Previously, we talked about changing a command's execFn to take no arguments by default, and changing `WithArguments` to accept a `func(...string) error`, which would be called back internally.
If it's not nil internally, it's implicit that the function takes arguments, and has provided a callback which handles the input arguments.
Depending on how args is passed around, this might be more or less complicated. Unsure at the moment.

**Comment by djdv** [comment]
We've been using this to build commands with and it seems to work well enough for now. It can always be refined later if needed.
master was rewritten (I forgot to put a license up front so I had to) with these incorporated into it rather than a direct merge.

**Pull request closed without merging**